File: solutions/cpp/0355.cpp
struct Tweet {
  int tweetId;
  int time;
  Tweet(int tweetId, int time) : tweetId(tweetId), time(time) {}
};

class Twitter {
 public:
  void postTweet(int userId, int tweetId) {
    tweets[userId].push_back(Tweet(tweetId, time++));
  }

  vector<int> getNewsFeed(int userId) {
    vector<int> newsFeed;

    auto compare = [](const Tweet& a, const Tweet& b) {
      return a.time < b.time;
    };
    priority_queue<Tweet, vector<Tweet>, decltype(compare)> pq(compare);

    // push your own tweets
    for (const Tweet& tweet : tweets[userId])
      pq.push(tweet);

    // push your followees' tweets
    for (const int followeeId : followees[userId])
      for (const Tweet& tweet : tweets[followeeId])
        pq.push(tweet);

    for (int i = min(10, (int)pq.size()); i > 0; --i)
      newsFeed.push_back(pq.top().tweetId), pq.pop();

    return newsFeed;
  }

  void follow(int followerId, int followeeId) {
    if (followerId == followeeId) return;
    followees[followerId].insert(followeeId);
  }

  void unfollow(int followerId, int followeeId) {
    if (followerId == followeeId) return;
    followees[followerId].erase(followeeId);
  }

 private:
  int time = 0;
  unordered_map<int, unordered_set<int>> followees;  // {userId: {followeeIds}}
  unordered_map<int, vector<Tweet>> tweets;          // {userId: {Tweets}}
};

File: solutions/java/0355.java
class Twitter {
  public void postTweet(int userId, int tweetId) {
    if (!tweets.containsKey(userId))
      tweets.put(userId, new LinkedList<>());
    tweets.get(userId).addFirst(new Pair<>(timer++, tweetId));
    if (tweets.get(userId).size() > 10)
      tweets.get(userId).removeLast();
  }

  public List<Integer> getNewsFeed(int userId) {
    List<Integer> ans = new ArrayList<>();
    Queue<Pair<Integer, Integer>> pq = new PriorityQueue<>((t1, t2) -> t2.getKey() - t1.getKey());
    for (Pair<Integer, Integer> tweet : tweets.getOrDefault(userId, new LinkedList<>()))
      pq.add(tweet);
    for (Integer followee : followees.getOrDefault(userId, new HashSet<>()))
      for (Pair<Integer, Integer> tweet : tweets.getOrDefault(followee, new LinkedList<>()))
        pq.add(tweet);
    for (int i = Math.min(10, pq.size()); i > 0; --i)
      ans.add(pq.poll().getValue());
    return ans;
  }

  public void follow(int followerId, int followeeId) {
    if (followerId == followeeId)
      return;
    if (!followees.containsKey(followerId))
      followees.put(followerId, new HashSet<>());
    followees.get(followerId).add(followeeId);
  }

  public void unfollow(int followerId, int followeeId) {
    if (followerId == followeeId)
      return;
    if (followees.containsKey(followerId))
      if (followees.get(followerId).contains(followeeId))
        followees.get(followerId).remove(followeeId);
  }

  private int timer = 0;
  private Map<Integer, Deque<Pair<Integer, Integer>>> tweets = new HashMap<>();
  private Map<Integer, Set<Integer>> followees = new HashMap<>();
}